File: src/main/java/org/goafabric/core/data/controller/PatientController.java
package org.goafabric.core.data.controller;

import jakarta.validation.Valid;
import jakarta.ws.rs.*;
import jakarta.ws.rs.core.MediaType;
import org.goafabric.core.data.controller.vo.Patient;
import org.goafabric.core.data.logic.PatientLogic;

import java.util.List;

@Path("/patients")
@Produces(MediaType.APPLICATION_JSON)
public class PatientController {
    private final PatientLogic logic;

    public PatientController(PatientLogic logic) {
        this.logic = logic;
    }

    @DELETE
    @Path("deleteById/{id}")
    public void deleteById(@PathParam("id") String id) {
        logic.deleteById(id);
    }

    @GET
    @Path("getById/{id}")
    public Patient getById(@PathParam("id") String id) {
        return logic.getById(id);
    }

    @GET
    @Path("findByGivenName")
    public List<Patient> findByGivenName(@QueryParam("givenName") String givenName) {
        return logic.findByGivenName(givenName);
    }

    @GET
    @Path("findByFamilyName")
    public List<Patient> findByFamilyName(@QueryParam("familyName") String familyName) {
        return logic.findByFamilyName(familyName);
    }

    @POST
    @Path("save")
    @Consumes(MediaType.APPLICATION_JSON)
    public Patient save(@Valid Patient patient) {
        return logic.save(patient);
    }

}

File: src/main/kotlin/org/goafabric/core/organization/controller/PatientController.kt
package org.goafabric.core.organization.controller

import jakarta.ws.rs.*
import jakarta.ws.rs.core.MediaType
import org.goafabric.core.organization.controller.dto.Patient
import org.goafabric.core.organization.logic.PatientLogic

@Path("/patients")
@Produces(MediaType.APPLICATION_JSON)
class PatientController(private val logic: PatientLogic) {

    @DELETE
    @Path("deleteById/{id}")
    fun deleteById(@PathParam("id") id: String) {
        logic.deleteById(id)
    }

    @GET
    @Path("getById/{id}")
    fun getById(@PathParam("id") id: String): Patient = logic.getById(id)

    @GET
    @Path("findByGivenName")
    fun findByGivenName(@QueryParam("givenName") givenName: String?): List<Patient> =
        logic.findByGivenName(givenName ?: "")

    @GET
    @Path("findByFamilyName")
    fun findByFamilyName(@QueryParam("familyName") familyName: String?): List<Patient> =
        logic.findByFamilyName(familyName ?: "")

    @GET
    @Path("findPatientNamesByFamilyName")
    fun findPatientNamesByFamilyName(@QueryParam("search") search: String?): List<Patient> =
        logic.findPatientNamesByFamilyName(search ?: "")

    @POST
    @Path("save")
    @Consumes(MediaType.APPLICATION_JSON)
    fun save(patient: Patient): Patient = logic.save(patient)
}
